controller/api/destination/endpoint_translator.go

```diff
@@ -609,3 +688,65 @@ func getInboundPort(podSpec *corev1.PodSpec) (uint32, error) {
 	}
 	return 0, fmt.Errorf("failed to find %s environment variable in any container for given pod spec", envInboundListenAddr)
 }
+
+// getInboundPortFromExternalWorkload gets the inbound port from the ExternalWorkload spec
+// variable.
+func getInboundPortFromExternalWorkload(ewSpec *ewv1alpha1.ExternalWorkloadSpec) (uint32, error) {
+	for _, p := range ewSpec.Ports {
+		if p.Name == pkgK8s.ProxyPortName {
+			return uint32(p.Port), nil
+		}
+	}
+
+	return 0, fmt.Errorf("failed to find %s port for given ExternalWorkloadSpec", pkgK8s.ProxyPortName)
+}
+
+// GetAnnotatedOpaquePortsForPod returns the opaque ports for the external workload given its
+// annotations, or the default opaque ports if it's not annotated
+func getAnnotatedOpaquePortsForExternalWorkload(ew *ewv1alpha1.ExternalWorkload, defaultPorts map[uint32]struct{}) map[uint32]struct{} {
+	if ew == nil {
+		return defaultPorts
+	}
+	annotation, ok := ew.Annotations[pkgK8s.ProxyOpaquePortsAnnotation]
+	if !ok {
+		return defaultPorts
+	}
+	opaquePorts := make(map[uint32]struct{})
+	if annotation != "" {
+		for _, pr := range parseExternalWorkloadOpaquePorts(annotation, ew) {
+			for _, port := range pr.Ports() {
+				opaquePorts[uint32(port)] = struct{}{}
+			}
+		}
+	}
+	return opaquePorts
+}
+
+func parseExternalWorkloadOpaquePorts(override string, ew *ewv1alpha1.ExternalWorkload) []util.PortRange {
+	portRanges := util.GetPortRanges(override)
+	var values []util.PortRange
+	for _, pr := range portRanges {
+		port, named := isNamedInExternalWorkload(pr, ew)
+		if named {
+			values = append(values, util.PortRange{UpperBound: int(port), LowerBound: int(port)})
+		} else {
+			pr, err := util.ParsePortRange(pr)
+			if err != nil {
+				log.Warnf("Invalid port range [%v]: %s", pr, err)
+				continue
+			}
+			values = append(values, pr)
+		}
+	}
+	return values
+}
+
+func isNamedInExternalWorkload(pr string, ew *ewv1alpha1.ExternalWorkload) (int32, bool) {
```

Comment:
Should we change the return type to an unsigned integer to be consistent with our types? (as an aside, I really don't get why k8s isn't using a 16-bit type for this...)

Reply:
Maybe, up to you really. Again, I am simply following patterns that are present in the rest of the code: 

in parsing.go we have `func isNamed(pr string, containers []corev1.Container) (int32, bool)`

Reply:
Let's leave as is, I've noticed more types using int32.